File: TwoWayDatabinding/app/src/main/java/com/enpassio/twowaydatabinding/viewmodel/AddToyViewModel.java
package com.enpassio.twowaydatabinding.viewmodel;

import android.arch.lifecycle.ViewModel;

import com.enpassio.twowaydatabinding.data.ToyRepository;
import com.enpassio.twowaydatabinding.data.model.Gender;
import com.enpassio.twowaydatabinding.data.model.ToyEntry;

import java.util.HashMap;
import java.util.Map;

public class AddToyViewModel extends ViewModel {

    private ToyRepository mRepo;
    private ToyEntry mChosenToy;
    private ToyEntry mToyBeingModified;
    private boolean mIsEdit;
    public static final String WOODEN = "Wooden";
    public static final String ELECTRONIC = "Electronic";
    public static final String PLASTIC = "Plastic";
    public static final String PLUSH = "Plush";
    public static final String MUSICAL = "Musical";
    public static final String EDUCATIVE = "Educative";

    public AddToyViewModel(ToyRepository repo, ToyEntry chosenToy) {
        mRepo = repo;
        if(chosenToy != null){
            //This is edit case
            mToyBeingModified = chosenToy.copy();
            mChosenToy = chosenToy;
            mIsEdit = true;
        } else {
            /*This is for adding a new toy. We initialize a ToyEntry with default or null values
            This is because two-way databinding in the AddToyFragment is designed to
            register changes automatically, but it will need a toy object to register those changes.*/
            mToyBeingModified = getEmptyToy();
            mIsEdit = false;
        }
    }

    public ToyEntry getToyBeingModified() {
        return mToyBeingModified;
    }

    private void insertToy(final ToyEntry toy){
        mRepo.insertToy(toy);
    }

    private void updateToy(final ToyEntry toy){
        mRepo.updateToy(toy);
    }

    public void saveToy(){
        if(!mIsEdit){
            insertToy(mToyBeingModified);
        } else {
            updateToy(mToyBeingModified);
        }
    }

    private ToyEntry getEmptyToy() {
        Map<String, Boolean> categories = new HashMap<>();
        categories.put(WOODEN, false);
        categories.put(ELECTRONIC, false);
        categories.put(PLASTIC, false);
        categories.put(PLUSH, false);
        categories.put(MUSICAL, false);
        categories.put(EDUCATIVE, false);
        return new ToyEntry("", categories, Gender.UNISEX, null);
    }

    public boolean isChanged(){
        boolean isChanged;
        if(mIsEdit){
            isChanged = !mToyBeingModified.equals(mChosenToy);
        } else {
            isChanged = !mToyBeingModified.equals(getEmptyToy());
        }
        return isChanged;
    }
}

File: TwoWayDatabindingKotlin/app/src/main/java/com/enpassion/twowaydatabindingkotlin/viewmodel/AddToyViewModel.kt
package com.enpassion.twowaydatabindingkotlin.viewmodel

import androidx.lifecycle.ViewModel
import com.enpassion.twowaydatabindingkotlin.data.ToyEntry
import com.enpassion.twowaydatabindingkotlin.data.ToyRepository


class AddToyViewModel(private val mRepo: ToyRepository, private val chosenToy: ToyEntry?) : ViewModel() {

    val toyBeingModified: ToyEntry

    private var mIsEdit: Boolean = false

    init {
        if (chosenToy != null) {
            //This is edit case
            toyBeingModified = chosenToy.copy()
            mIsEdit = true
        } else {
            /*This is for adding a new toy. We initialize a ToyEntry with default or null values
            This is because two-way databinding in the AddToyFragment is designed to
            register changes automatically, but it will need a toy object to register those changes.*/
            toyBeingModified = emptyToy
            mIsEdit = false
        }
    }

    private fun insertToy(toy: ToyEntry) {
        mRepo.insertToy(toy)
    }

    private fun updateToy(toy: ToyEntry) {
        mRepo.updateToy(toy)
    }

    fun saveToy() {
        if (!mIsEdit) {
            insertToy(toyBeingModified)
        } else {
            updateToy(toyBeingModified)
        }
    }

    private val emptyToy: ToyEntry
        get() {
            val categories = mutableMapOf(
                WOODEN to false, ELECTRONIC to false,
                PLASTIC to false, PLUSH to false, MUSICAL to false, EDUCATIVE to false
            )
            return ToyEntry(toyName = "", categories = categories)
        }


   var isChanged : Boolean = false
        get() = if(mIsEdit) toyBeingModified != chosenToy
                    else toyBeingModified != emptyToy
        private set

    companion object {
        const val WOODEN = "Wooden"
        const val ELECTRONIC = "Electronic"
        const val PLASTIC = "Plastic"
        const val PLUSH = "Plush"
        const val MUSICAL = "Musical"
        const val EDUCATIVE = "Educative"
    }
}